### ISSUE TITLE
[STREAMING] Replace stub tensor conversion functions with real implementations in streaming.rs

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#328]
## Problem Description

The streaming generation module in `crates/bitnet-inference/src/streaming.rs` contains critical stub implementations for tensor conversion functions that are blocking real neural network inference. The `tokens_to_tensor` and `tensor_to_logits` functions currently return mock data instead of performing actual tensor operations, preventing proper integration with the BitNet quantization and inference pipeline.

## Environment Details

**File:** `crates/bitnet-inference/src/streaming.rs`  
**Functions:** `tokens_to_tensor` (line 445), `tensor_to_logits` (line 450)  
**Codebase:** BitNet.rs neural network inference  
**Feature Flags:** Affects both `cpu` and `gpu` feature compilation paths  

### Current Stub Implementations

```rust
/// Convert tokens to input tensor (mock implementation)
fn tokens_to_tensor(tokens: &[u32]) -> Result<ConcreteTensor> {
    Ok(ConcreteTensor::mock(vec![1, tokens.len()]))
}

/// Extract logits from output tensor (mock implementation)
fn tensor_to_logits(_tensor: &ConcreteTensor, vocab_size: usize) -> Result<Vec<f32>> {
    // In a real implementation, this would extract logits from the tensor
    Ok(vec![0.1; vocab_size])
}
```

### Technical Issues

1. **Mock Tensor Creation**: `tokens_to_tensor` creates mock tensors instead of converting u32 token IDs to real tensors that can be processed by quantized linear layers
2. **Dummy Logits Generation**: `tensor_to_logits` returns uniform dummy values (0.1) instead of extracting actual logits from model output tensors
3. **Backend Integration Gap**: These stubs break the integration with `Backend::forward()` implementations which expect real `ConcreteTensor` data
4. **Streaming Pipeline Failure**: The entire streaming generation system (`GenerationStream`) produces meaningless output due to these mock implementations

## Impact Assessment

**Severity:** High - Blocking core functionality  
**Affected Components:**
- Streaming token generation (`GenerationStream`)
- Server-side streaming endpoints (`bitnet-server`)
- Real-time inference applications  
- Production deployment readiness

**User Impact:**
- Streaming generation returns meaningless repeated tokens
- Integration tests may pass but produce invalid results  
- Production inference servers cannot generate coherent text

### Primary Implementation Approach

Replace stub functions with real tensor operations that integrate with the existing `ConcreteTensor` enum and backend infrastructure:

```rust
/// Convert tokens to input tensor for model processing
fn tokens_to_tensor(tokens: &[u32]) -> Result<ConcreteTensor> {
    if tokens.is_empty() {
        return Err(anyhow::anyhow!("Cannot create tensor from empty token sequence"));
    }
    
    // Create tensor depending on available backend
    #[cfg(feature = "gpu")]
    {
        use candle_core::{Device, Tensor};
        let device = Device::cuda_if_available(0)?;
        let tensor = Tensor::new(tokens, &device)?
            .unsqueeze(0)?; // Add batch dimension
        Ok(ConcreteTensor::BitNet(tensor))
    }
    
    #[cfg(not(feature = "gpu"))]
    {
        use candle_core::{Device, Tensor};
        let tensor = Tensor::new(tokens, &Device::Cpu)?
            .unsqueeze(0)?; // Add batch dimension  
        Ok(ConcreteTensor::BitNet(tensor))
    }
}

/// Extract logits from model output tensor
fn tensor_to_logits(tensor: &ConcreteTensor, vocab_size: usize) -> Result<Vec<f32>> {
    match tensor {
        ConcreteTensor::BitNet(candle_tensor) => {
            // Handle different tensor shapes from model outputs
            let shape = candle_tensor.shape();
            let logits_tensor = if shape.rank() == 3 {
                // Shape: [batch_size, seq_len, vocab_size] -> take last token
                candle_tensor.i((0, shape.dims()[1] - 1))?
            } else if shape.rank() == 2 {
                // Shape: [batch_size, vocab_size] -> take batch 0
                candle_tensor.i(0)?
            } else {
                return Err(anyhow::anyhow!(
                    "Unexpected tensor shape for logits extraction: {:?}. Expected rank 2 or 3.",
                    shape
                ));
            };
            
            // Validate vocabulary size
            let actual_vocab_size = logits_tensor.dim(0)?;
            if actual_vocab_size != vocab_size {
                return Err(anyhow::anyhow!(
                    "Vocabulary size mismatch: expected {}, got {}",
                    vocab_size, actual_vocab_size
                ));
            }
            
            // Convert to Vec<f32>
            let logits = logits_tensor.to_vec1::<f32>()?;
            Ok(logits)
        },
        ConcreteTensor::Mock(mock_tensor) => {
            // Fallback for test environments
            if mock_tensor.shape().len() >= 2 {
                let vocab_size = mock_tensor.shape()[mock_tensor.shape().len() - 1];
                Ok(vec![0.1; vocab_size])
            } else {
                Ok(vec![0.1; vocab_size])
            }
        }
    }
}
```

### Alternative Approaches

1. **Backend-Delegated Conversion**: Move tensor conversion logic into `Backend` trait implementations
2. **Device-Specific Optimizations**: Create specialized conversion paths for CPU (SIMD) vs GPU (CUDA) tensors
3. **Lazy Tensor Creation**: Use tensor builders for more efficient memory management

### Phase 1: Core Implementation (Priority: High)
- [ ] Replace `tokens_to_tensor` with real tensor creation logic
- [ ] Replace `tensor_to_logits` with proper tensor extraction
- [ ] Add comprehensive error handling for tensor operations
- [ ] Ensure feature flag compatibility (`cpu`/`gpu` builds)

### Phase 2: Integration & Testing (Priority: High)  
- [ ] Update `forward_pass` function to handle real tensor data
- [ ] Verify integration with existing `Backend::forward()` implementations
- [ ] Add tensor shape validation and error recovery
- [ ] Test with different model architectures and vocab sizes

### Phase 3: Optimization (Priority: Medium)
- [ ] Optimize tensor creation for streaming performance
- [ ] Add device-aware tensor allocation (CPU vs GPU)
- [ ] Implement memory pooling for repeated tensor operations
- [ ] Profile and optimize tensor conversion overhead

### Phase 4: Validation (Priority: High)
- [ ] Create integration tests with real model tensors
- [ ] Validate streaming output against non-streaming inference
- [ ] Test with various quantization types (I2_S, TL1, TL2)
- [ ] Cross-validate with C++ reference implementation

### Unit Tests
```rust
#[test]
fn test_tokens_to_tensor_shape() {
    let tokens = vec![1, 2, 3, 4, 5];
    let tensor = tokens_to_tensor(&tokens).unwrap();
    match tensor {
        ConcreteTensor::BitNet(t) => {
            assert_eq!(t.shape().dims(), &[1, 5]); // [batch_size, seq_len]
        },
        _ => panic!("Expected BitNet tensor")
    }
}

#[test] 
fn test_tensor_to_logits_extraction() {
    // Create mock model output tensor [1, 1, vocab_size]
    let mock_logits = vec![0.1, 0.8, 0.05, 0.05]; // Example vocab_size=4
    let tensor = create_test_tensor(vec![1, 1, 4], mock_logits);
    let logits = tensor_to_logits(&tensor, 4).unwrap();
    assert_eq!(logits.len(), 4);
    assert!((logits[1] - 0.8).abs() < 1e-6); // Check extraction accuracy
}
```

### Integration Tests
- End-to-end streaming with real model weights
- Performance comparison: streaming vs non-streaming inference
- Memory usage validation during long generation sequences
- Error handling with malformed tensors and invalid token sequences

### Potential Complications
1. **Tensor Shape Mismatches**: Different model architectures may produce varying output shapes
2. **Device Memory Issues**: GPU tensor allocation failures during streaming
3. **Performance Regression**: Real tensor operations may introduce latency
4. **Quantization Compatibility**: Ensuring compatibility with I2_S/TL1/TL2 quantized tensors

### Mitigation Strategies
1. **Robust Shape Handling**: Support multiple tensor shapes with fallback logic
2. **Memory Management**: Implement tensor pooling and cleanup strategies  
3. **Performance Monitoring**: Add metrics to track tensor conversion overhead
4. **Quantization Testing**: Validate with all supported quantization methods

## Related Issues and Dependencies

**Dependencies:**
- Requires functional `ConcreteTensor` implementations in `bitnet-common`
- Depends on `candle-core` tensor operations  
- Needs working `Backend::forward()` implementations

**Related Issues:**
- Issue #260: Real quantized neural network inference implementation
- Streaming generation accuracy validation
- Production inference server architecture (Issue #251)

### Definition of Done
- [ ] `tokens_to_tensor` creates valid tensors from token sequences
- [ ] `tensor_to_logits` extracts correct logits from model outputs  
- [ ] Streaming generation produces coherent, meaningful text
- [ ] Integration tests pass with real model weights
- [ ] Performance overhead is within acceptable limits (<10% latency increase)
- [ ] Error handling covers edge cases (empty tokens, shape mismatches)
- [ ] Feature flag compatibility maintained (`cpu`/`gpu` builds)

### Success Metrics
- Streaming generation produces text with semantic coherence
- Token-by-token output matches non-streaming inference results
- Memory usage remains stable during long generation sequences
- Integration with existing quantization and backend systems works correctly

## Implementation Notes

**Code Location:** `/home/steven/code/Rust/BitNet-rs/crates/bitnet-inference/src/streaming.rs`  
**Key Functions:** Lines 445-453 (`tokens_to_tensor`, `tensor_to_logits`)  
**Testing:** Update tests in `streaming.rs` test module and add integration tests  
**Documentation:** Update function documentation to reflect real implementations

This issue blocks the streaming generation pipeline and is essential for production deployment of the BitNet.rs inference server. The implementation should follow the existing codebase patterns and maintain compatibility with the feature flag system.
